StableHLO Versioned Dialect and Compatibility Passes

GleasonK opened the pull request:
Add StableHLO compatibility dialect and passes for reading and writing with forward/backward compatibility guarantees.

**Note: This is still a prototype implementation and should not be used in production until RFCs have been approved and types have been forked.**

```bash
stablehlo-opt [flags] file.mlir
    --stablehlo-legalize-to-vhlo 
    --vhlo-to-version='target=[version]'    Translate versioned dialect to target version (current, 0.3.0, ...)
    --vhlo-legalize-to-stablehlo 
```

Change description:
- Introduce VHLO, the Versioned StableHLO Dialect.
  + This dialect is a shallow copy of StableHLO's in-memory layout. It does not include verifiers or constraints.
  + Once an op is added to VHLO it must remain unchanged so that it can be guaranteed that a VHLO op is identical across versions.
  + The first version of VHLO is `0.3.0`.
- Conversion passes for compatibility
  + StableHLO <--> VHLO legalizations. StableHLO is always able to be legalized to/from the latest version of VHLO.
  + VHLO-to-version. Target previous versions of VHLO ops for forward compatibility. Upgrade to the latest version of VHLO ops to emit StableHLO.
- Testing for legalizations and version conversions.

Future work (these items will be made into individual GH issues before submit):
- Think more about a scalable way to test this as StableHLO evolves.
- Additional feature work on the tool.. any missing flags? Pass pipeline for simplicity?
- Improve user experience.
- See open [compatibility issues](https://github.com/openxla/stablehlo/labels/Compatibility) 

Closes #255

burmako commented on stablehlo/tests/compat_downgrade.mlir:
```diff
@@ -0,0 +1,10 @@
+// RUN: stablehlo-opt --stablehlo-legalize-to-versionedhlo --versionedhlo-to-version='target=minimum' %s | FileCheck %s
```
Let's have separate tests for separate passes to improve coverage.

For testing the legalization passes, we can outright copy/paste https://github.com/tensorflow/mlir-hlo/blob/master/tests/Dialect/mhlo/stablehlo-legalize-to-hlo.mlir into `stablehlo-legalize-to-vhlo.mlir` and adjust as needed. (Same for the other direction of the legalization).

Testing to-version is trickier:
  * I'd propose having one test file per version, i.e. `vhlo-to-version-v0.3.0.mlir`, `vhlo-to-version-v0.3.1.mlir`, etc, and each of those files would test `vhlo-to-version=v0.3.0`, `vhlo-to-version=v0.3.1`, etc (these tests would exercise only the ops which changed in this particular version, not the entire opset). 
  * Then we could have `vhlo-to-version.mlir` that tests generic machinery like the `invalid_program_unknown_op` below.
  * I don't think we should test `minimum` or `current`, since they are just aliases to concrete versions, and I don't think that we should burden humans with remembering which versions these aliases are referring to.

GleasonK replied:
A few thoughts:
- Opset changes should be accompanied by minor bumps. Given that, maybe we name them `vhlo-to-version-v0.3.mlir`, etc.
- What do we do for opset bumps when new ops are introduced? `vhlo-to-version-0.4.mlir` would have a single negative test?
- I like the idea of having separate test files for the sake of exhaustively testing each upgrade/downgrade, but that seems like it would accumulate over time, let's discuss _what_ versions are. Should we do a single change to a single version, or if we have 5 PR's ready with modifications, do some sort of batching?
- I agree with most of this, aside from _"I don't think we should test `current`"_. This is actually the entirety of how LLVM tests backward compatibility:
  + ([LLVM Backward Compat Policy](https://llvm.org/docs/DeveloperPolicy.html#ir-backwards-compatibility)
  + [LLVM Compatibility Test Files `compatibility.*`](https://github.com/llvm/llvm-project/tree/304f1d59ca41872c094def3aee0a8689df6aa398/llvm/test/Bitcode)
  + [Example of changes to a compat test over time](https://github.com/llvm/llvm-project/commits/304f1d59ca41872c094def3aee0a8689df6aa398/llvm/test/Bitcode/compatibility-3.8.ll). Note that filecheck's are allowed to change over time but the IR is not.


All-in-all, I propose that we separate this into another ticket that we address in the near future, I think this needs a bit more discussion and doesn't need to preclude submission of this CL. The tests in this CL are sufficient for where we stand right now, but I agree they don't scale well.

GleasonK replied:
"For testing the legalization passes"

I added a legalization pass, then it occurred to me that we probably don't need separate legalization tests. Given the 1:1 guaranteed parity, we would have identical test files that effectively accomplish what a round trip test would. So I made a `legalize_stablehlo_to_vhlo.mlir` file, and tested using FileCheck to get to VHLO, then tested round-trip StableHLO->VHLO->StableHLO to ensure that VHLO->StableHLO is working as expected.

burmako commented on stablehlo/tests/compat_downgrade.mlir:
```diff
@@ -0,0 +1,10 @@
+// RUN: stablehlo-opt --stablehlo-legalize-to-versionedhlo --versionedhlo-to-version='target=minimum' %s | FileCheck %s
```
Speaking of aliases, what's the use case for having the `minimum` alias? 

I can see how `current` is useful. When deserializing StableHLO programs, I'd want to run `vhlo-to-version=current` without thinking what exactly is the version I want, and then I'd want to `vhlo-legalize-to-stablehlo`.

What are the use cases where someone would want to indiscriminately downgrade to `minimum`, regardless of what version that actually is?

GleasonK replied:
Intention was for a user who wouldn't want to have to continue updating the minimum supported target version. I.e. producer is on 0.5.x and some consumers are running a slightly different versions depending on their update cycle 0.2.x, 0.6.x. Targeting minimum is seamless and safe. The alternative would be knowing exactly what the minimum consumer number is and targeting that, and then updating that number any time the minimum consumer number changes.

That said, it probably isn't that useful to target minimum. Especially if producers want to start using new features that aren't in the minimum supported version, then they will need to go to the above system where they target the minimum (or less than minimum) consumer version.

I agree we should remove this. Tests can be updated to `0.0.0`.

burmako commented on stablehlo/compatibility/TODO.txt:
```diff
@@ -0,0 +1,9 @@
+4. Validate target version string.
+5. Add versioned dialect type forks.
+6. Add bytecode for Versioned StableHLO
+7. Add versioned CHLO ops
```
Versioned CHLO is the lowest priority of these work items - we can push it into 2023.

GleasonK replied:
Sounds good. Will note that in the GH issue.

burmako reviewed (changes requested):
Monumental work! I left a bunch of comments, but I think we're aligned on the high-level design, so the comments are all about naming and small-scale details.

Let's land this asap and iterate on completing the prototype in smaller PRs. No need to wait for the review of #115, since the PR doesn't touch the StableHLO opset.

GleasonK commented on stablehlo/compatibility/dialect/CMakeLists.txt:
```diff
@@ -0,0 +1,50 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set(LLVM_TARGET_DEFINITIONS VersionedStablehloOps.td)
```
Done. Just pushed two commits that address this one comment. Will get to the rest now.

GleasonK commented on stablehlo/compatibility/transforms/CompatibilityPasses.cpp:
```diff
@@ -0,0 +1,621 @@
+      return emitToVersionError(op,
+                                "op has a non-empty result_layouts attribute");
+    }
+    if (op->hasAttr("result_layout")) op->removeAttr("result_layout");
```
Fixed. I wasn't sure if `(getResultLayout() -> None) == !hasAttr(result_layout)`. I.e if there was a way that something like `result_layout = []` would return none, but the attribute is there.

Will delete this line.

GleasonK commented on stablehlo/compatibility/transforms/CompatibilityTypeConversion.h:
```diff
@@ -0,0 +1,66 @@
+      if (failed(convertTypes(type.getTypes(), convertedTypes))) return {};
+      return TupleType::get(type.getContext(), convertedTypes);
+    });
+    // FIXME: Do I need to do anything with ranked tensor / encoding here?
```
Will make a GH issue for all FIXME's and TODO's left.

GleasonK commented on stablehlo/compatibility/dialect/VersionNumber.h:
```diff
@@ -0,0 +1,99 @@
+namespace mlir {
+namespace versionedhlo {
+
+class VersionNumber {
```
I went back and forth on this naming. I think `Version` is probably fine, no real opinion either way. I'll make that change.

GleasonK commented on stablehlo/compatibility/dialect/VersionedStablehloAttrs.td:
```diff
@@ -0,0 +1,174 @@
+  let assemblyFormat = "`<` struct(params) `>`";
+}
+
+def OutputOperandAlias : AttrDef<VersionedHLO_Dialect, "OutputOperandAlias"> {
```
Huh. This happened because it is called `OutputOperandAlias` in `StablehloAttrs.td` instead of `StableHLO_OutputOperandAlias`. I'll make that fix too.

GleasonK commented on stablehlo/compatibility/dialect/VersionedStablehloBase.td:
```diff
@@ -0,0 +1,182 @@
+    
+    Version log:
+      0.0.0: Base release.
+      0.1.0: CustomCallOp add result aliasing.
```
I didn't know internally we have a name for the base release. The proposal in the Compat RFC is that all changes to the VHLO opset are accompanied by a minor version bump, since they have backward/forward compatibility implications, more than a patch. Patch bumps would be reserved for compat-exempt fixes to documentation or verifiers.

Also I agree that we should do this in a followup ticket. We could discuss batching there as well:

```
    Version log:
      0.3.0: Bootstrap from MHLO: https://github.com/openxla/stablehlo/pull/1.
      0.4.0: Add AllGatherOp::use_global_device_ids: https://github.com/openxla/stablehlo/pull/272.
             Add CollectivePermuteOp::channel_id: https://github.com/openxla/stablehlo/pull/388.
             Add CustomCallOp::output_operand_aliases: https://github.com/openxla/stablehlo/pull/403.
```

GleasonK replied:
I modified this to include PR links, but am going to wait to call the base release 0.3.0 until we align on what these numbers mean. I'm not sure the internal numbering scheme aligns with the external release numbering proposed in the Compat RFC.

GleasonK replied:
Sounds good, will update to use this numbering scheme.

GleasonK commented on stablehlo/compatibility/dialect/VersionedStablehloAttrs.td:
```diff
@@ -0,0 +1,174 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+   Copyright 2022 The VersionedHLO Authors.
```
Yeah...sed scripting at its finest here.
Fixed.

GleasonK commented on stablehlo/compatibility/dialect/VersionedStablehloBase.td:
```diff
@@ -0,0 +1,182 @@
+// Versioned Dialect Constants and Declaration
+//===----------------------------------------------------------------------===//
+
+def VersionedHLO_Dialect : Dialect {
```
Can fix that. Moved because the dependency structure seemed strange, and I encountered build errors when I was getting started on this work. `*Enums.td` and `*Attrs.td` requires the definition of the Dialect, and are included by `*Ops.td` where the Dialect is defined. Base felt like a fitting place to put the definitions that all other `Vhlo*.td` files needed.

StableHLO and CHLO get around this by including these two files after a few lines of declarations in the tablegen, like a `.inc` include, which LLVM coding style would denote as "[hidden dependencies](https://llvm.org/docs/CodingStandards.html#include-style)" (for C++ at least, maybe this is common practice for tablegen).

GleasonK commented on stablehlo/compatibility/dialect/VersionedStablehloAttrs.td:
```diff
@@ -0,0 +1,174 @@
+  let assemblyFormat = "`<` struct(params) `>`";
+}
+
+def OutputOperandAlias : AttrDef<VersionedHLO_Dialect, "OutputOperandAlias"> {
```
I think I'm going to delete `OutputOperandAlias` all together. Let's address this in a follow-up CL with the other changes mentioned in a separate comment.

GleasonK commented on stablehlo/compatibility/dialect/VersionedStablehloOps.cpp:
```diff
@@ -0,0 +1,95 @@
+    os << "token";
+    return;
+  }
+  os << "<unknown stablehlo type>";
```
fixed here and two other occurrences, for attr/op/type.

GleasonK commented on stablehlo/compatibility/dialect/VersionedStablehloOps.td:
```diff
@@ -0,0 +1,976 @@
+    InterfaceMethod<
+      "Description...", "mlir::versionedhlo::VersionNumber", "getMaxVersion">,
+  ];
+}
```
Probably should. The misplacement is also an indication that attrs/types are not versioned currently. Added that to TODO.txt while I'm fixing up, will make a GH issue before submit.

Moved to VhloBase.td

GleasonK commented on stablehlo/compatibility/dialect/VersionedStablehloOps.h:
```diff
@@ -0,0 +1,83 @@
+
+  static llvm::StringRef getMinimumDialectVersion() {
+    return "0.0.0";
+  }
```
This was made before `VersionNumber` was a thing. I agree, I think a better design would be to have Ops return a `pair<VersionNumber, VersionNumber>` which denotes min-max dialect versions that a given op is supported in.

GleasonK replied:
Reworked. `VersionNumber` has no concept of current or minimum version. Only `VhloDialect` knows those details.

GleasonK commented on stablehlo/compatibility/dialect/VersionedStablehloBase.td:
```diff
@@ -0,0 +1,182 @@
+#ifndef STABLEHLO_COMPATIBILITY_VERSIONEDHLO_BASE
+#define STABLEHLO_COMPATIBILITY_VERSIONEDHLO_BASE
+
+include "mlir/IR/OpBase.td"
```
Removed unnecessary includes, largely surrounding Interfaces:

```
// Used for AttrInterfaces, not needed in VHLO
include "mlir/IR/AttrTypeBase.td"
include "mlir/Interfaces/InferTypeOpInterface.td"

// Can be deleted from both. 
// No reference is made to ODS defs, only cpp quant
// references which are needed in StableHLO/VHLO cpp files.
include "mlir/Dialect/Quant/QuantOpsBase.td"
```

Deleted QuantOpsBase include to keep these files _somewhat_ more consistent.

GleasonK commented on stablehlo/compatibility/dialect/VersionNumber.h:
```diff
@@ -0,0 +1,99 @@
+    static VersionNumber getCurrent() {
+      return VersionNumber(CURRENT_VERSION);
+    }
+    static VersionNumber getMinimumSupported() {
```
These are deleted. Will add documentation to `VhloDialect::getCurrentDialectVerison()` / `VhloDialect::getMinimumDialectVersion()`.

GleasonK commented on stablehlo/compatibility/dialect/VersionNumber.h:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    std::string version;
+};
```
This cleaned things up a ton. Went with an approach where we store a `std::array<int64_t, 3>` so we can leverage array comparison operators. I.e. `std::array::operator<` uses `lexicographical_compare` by default which is exactly what we want for version numbers.

GleasonK replied:
Fixed so that all public aspects of this class are individual fields, array is an implementation detail now.

Keeping member variable  since comparisons are common and the options are either to re-implement a 3-field lexicographical compare, or construct a container to use for the std algorithm on each compare.

GleasonK commented on stablehlo/compatibility/dialect/VersionedStablehloAttrs.td:
```diff
@@ -0,0 +1,174 @@
+      VersionedHLO_Dim:$scatterDimsToOperandDims,
+      "int64_t":$indexVectorDim
+  );
+  let assemblyFormat = "`<` struct(params) `>`";
```
Want to avoid custom assembly if necessary. VHLO isn't really supposed to be kept in textual form (since textual assembly isn't stable) and ideally VHLO should be treated as an implementation detail. 

My goal was to keep as much of the versioned dialect in ODS as possible for maintainability. However when an attr has a mnemonic, some form of custom assembly must be specified, and this was a minimal option.

I've added a task for "figure out what to do for custom assembly in VHLO". There is no harm in sharing assembly formats, since we offer no stability guarantees on textual assembly. But this feels like an area of "if we offer it, someone will rely on it" since textual assembly is _relatively_ stable.

GleasonK commented on stablehlo/compatibility/dialect/VersionNumber.h:
```diff
@@ -0,0 +1,99 @@
+    static VersionNumber getCurrent() {
+      return VersionNumber(CURRENT_VERSION);
+    }
+    static VersionNumber getMinimumSupported() {
```
These methods are deleted in favor of `VhloDialect::getMinimumDialectVersion()` / `VhloDialect::getCurrentDialectVersion()`

GleasonK commented on stablehlo/compatibility/TODO.txt:
```diff
@@ -0,0 +1,9 @@
+This file is not to be submitted. Just needed to log the work still to be done somewhere central.
```
Agree. Will create a separate issue for each item that remain in TODO, then will delete this file before submit.

GleasonK commented on stablehlo/CMakeLists.txt:
```diff
@@ -12,6 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+add_subdirectory(compatibility)
```
I moved the VHLO dialect to `stablehlo/dialect`. Unsure if transforms fit in with the dialect folder as well. The reason I chose a separate directory in the first place is that compatibility infrastructure feels like a feature of the StableHLO opset the same way that a reference interpreter is a feature of StableHLO, separate from the dialect.

Let's talk about this a bit more, I'm going to address the rest of the comments and leave this open for now. Currently this is 12 files that would be added to `stablehlo/dialect` that are all related to compatibility infrastructure. Also part of me wants to split `CompatibilityPasses` into separate files in the near future, since the relation between legalizations to/from VHLO and the VHLO-to-version conversions are conceptually different.

GleasonK replied:
Followed up offline. Directory structure will be:

```
stablehlo/
    dialect/
        Vhlo*
        Chlo*
        Stablehlo*
    transforms/
        LegalizeStablehloToVhlo*
        VhloToVersion*

```

GleasonK commented on stablehlo/compatibility/dialect/Register.cpp:
```diff
@@ -0,0 +1,29 @@
+namespace versionedhlo {
+
+void registerAllDialects(mlir::DialectRegistry &registry) {
+  registry.insert<mlir::versionedhlo::VersionedhloDialect>();
```
Agree. No need for this extra file. Deleted these register files and moved into registerAllDialects.

GleasonK commented on stablehlo/compatibility/dialect/VersionedStablehloOps.td:
```diff
@@ -0,0 +1,976 @@
+  let assemblyFormat = "$results attr-dict (`:` type($results)^)?";
+}
+
+///// SPLIT //////
```
Should have been a `FIXME: Remove`. I added ops a bit at a time, moving this marker down in the file to easily get to the spot. This was the last spot where I then added all below the line and forgot to clean up.

Deleted.

GleasonK commented on stablehlo/compatibility/transforms/CompatibilityPasses.cpp:
```diff
@@ -0,0 +1,621 @@
+
```
Hah - I left this exact reply on a previous comment. Will split. Still want to consider if these all belong in `stablehlo/dialect` or not.

GleasonK replied:
Split `Passes.cpp` into:

```
LegalizeStablehloToVhlo.cpp
LegalizeVhloToStablehlo.cpp
LegalizeVhloToVersion.cpp
```

I still have `Passes.h` with each cpp file implementing a different populate method, should I split that into separate legalize headers as well?

GleasonK commented on stablehlo/compatibility/transforms/CompatibilityPasses.td:
```diff
@@ -0,0 +1,35 @@
+  let summary = "Convert between versions of VersionedHLO.";
+  let options = [
+    Option<"targetVersion", "target", "std::string", "\"current\"",
+           "The target version for downgrade. (Defaults to minimum supported)">,
```
I wasn't able to find a way to not have a default. Could use an invalid number and check in the code.

GleasonK replied:
Updated to empty string default value and updated error checking to display the following:

```
$ stablehlo-opt stablehlo/tests/compat_upgrade.mlir --vhlo-to-version
stablehlo/tests/compat_upgrade.mlir:0:0: error: No target version specified. Specify target using: --vhlo-to-version='target=[targetVersion]'
Target version must be of the form #.#.# or 'current'.

$ stablehlo-opt stablehlo/tests/compat_upgrade.mlir --vhlo-to-version='target=x.y.z'
stablehlo/tests/compat_upgrade.mlir:0:0: error: Invalid target version number argument 'x.y.z'
Target version must be of the form #.#.# or 'current'.
```

GleasonK replied:
Is there a good way to unit test these negative test cases? I _could_ make an empty file for each and use verify diagnostics.. but it feels like there's probably a better way?

GleasonK replied:
Added separate test files. Was only 4 tests, not too bad.

GleasonK commented on stablehlo/compatibility/transforms/CompatibilityPasses.cpp:
```diff
@@ -0,0 +1,621 @@
+/////////////////////////////////////////
+
+// versionedhlo.custom_call --> versionedhlo.custom_call_v2
+struct CustomCallOpV2Upgrade
```
Good point.

GleasonK commented on stablehlo/compatibility/dialect/VersionedStablehloOps.td:
```diff
@@ -0,0 +1,976 @@
+
+def VersionedHLO_CreateTokenOp : VersionedHLO_Op<"create_token"> {
+  let results = (outs VHLO_Token:$output);
+}
```
Will align them. For the most part is the same. I just picked a few easy ones to test first, and I think those are at the top now.

GleasonK commented on stablehlo/compatibility/transforms/CompatibilityPasses.h:
```diff
@@ -0,0 +1,48 @@
+#include "stablehlo/compatibility/transforms/CompatibilityPasses.h.inc"
+
+/// Registers all transformation passes.
+void registerStablehloCompatibilityPasses();
```
Deleted to make use of the generated function.

Should pass registration be a part of the StableHLO `Register.cpp`? Or is that intended for Dialect registration only?

GleasonK commented on stablehlo/compatibility/transforms/MapStablehloToVersionedStablehlo.h:
```diff
@@ -0,0 +1,176 @@
+MAP_STABLEHLO_TO_VERSION_V0(UniformDequantizeOp)
+MAP_STABLEHLO_TO_VERSION_V0(UniformQuantizeOp)
+MAP_STABLEHLO_TO_VERSION_V0(WhileOp)
+MAP_STABLEHLO_TO_VERSION_V0(XorOp)
```
Done. This required updating `VhloOps.td` to suffix each op with `V1`, which is reasonable and makes the transformation code more clear.

GleasonK commented on stablehlo/compatibility/transforms/CompatibilityPasses.cpp:
```diff
@@ -0,0 +1,621 @@
+/// Registers all Torch transformation passes.
+void registerStablehloCompatibilityPasses() { registerPasses(); }
+
+// From CompatibilityTypeConversion.h
```
Created `TypeConversion.cpp`

GleasonK commented on stablehlo/compatibility/transforms/CompatibilityPasses.cpp:
```diff
@@ -0,0 +1,621 @@
+////////////////////////////////////////////
+
+LogicalResult emitToVersionError(Operation* op, llvm::StringRef message) {
+  return op->emitError("failed to downgrade ")
```
Is the concern that this applies to `ToVersion` in both directions? 

This is only used for downgrades since upgrades should never fail. Renamed the function to `emitDowngradeError`.

GleasonK commented on stablehlo/compatibility/transforms/CompatibilityPasses.cpp:
```diff
@@ -0,0 +1,621 @@
+    }
+    return success();
+  }
+};
```
I structured it this way because the majority of upgrades (historically) will be for default values attrs which can call the default builder methods. Structuring this way allows for very simple code, and taking it a step further, I had imagined making success the default implementation for prepareOpForConversion, meaning we can write most of our upgrades as:

Currently, I think we can accommodate all upgrades in the PR queue using:

```
struct CustomCallOpV1ToV2
    : public VersionConversionPattern<CustomCallOpV1, CustomCallOpV2> {
  using VersionConversionPattern::VersionConversionPattern;
};

struct CustomCallOpV2ToV3
    : public VersionConversionPattern<CustomCallOpV2, CustomCallOpV3> {
  using VersionConversionPattern::VersionConversionPattern;
};

struct AllGatherOpV1ToV2
    : public VersionConversionPattern<AllGatherOpV1, AllGatherOpV2> {
  using VersionConversionPattern::VersionConversionPattern;
};

struct CollectivePermuteOpV1ToV2
    : public VersionConversionPattern<CollectivePermuteOpV1, CollectivePermuteOpV2> {
  using VersionConversionPattern::VersionConversionPattern;
};
```

Moves/Renames can be modeled with:

```
// Upgrade
struct OriginalNameToNewName
    : public VersionConversionPattern<OriginalName, NewName> {
  using VersionConversionPattern::VersionConversionPattern;
};

// Downgrade
struct NewNameToOriginalName
    : public VersionConversionPattern<NewName, OriginalName> {
  using VersionConversionPattern::VersionConversionPattern;
};
```

Attribute renames and new attrs that don't have a default value can be fixed up using prepareOpForConversion. I'd almost go as far as marking `matchAndRewrite` as `final`.

I suppose we could also go the free function route and leverage a convenience macro as well to build these super-common cases. That would be similar to how IREE does it IIRC.

GleasonK commented on stablehlo/compatibility/transforms/CompatibilityPasses.cpp:
```diff
@@ -0,0 +1,621 @@
+struct CustomCallOpV2Upgrade
+    : public VersionConversionPattern<CustomCallOp, CustomCallOpV2> {
+  using VersionConversionPattern<CustomCallOp,
+                                 CustomCallOpV2>::VersionConversionPattern;
```
Done.

GleasonK commented on stablehlo/compatibility/dialect/VersionedStablehloOps.td:
```diff
@@ -0,0 +1,976 @@
+}
+
+class VersionedHLO_Op<string mnemonic, string minVersion = "0.0.0", string maxVersion = "current"> :
+    VersionedHLO_OpWithTraits<mnemonic, [], minVersion, maxVersion>;
```
My thoughts:
1. It is more difficult to misuse this way, must deliberately choose this class for traits.
2. There is some slight code-ugliness if we don't in that every op with a non-default version will need to specify empty trait list:

```
def VHLO_CustomCallOpV1 : VHLO_Op<"custom_call", [], "0.0.0", "0.0.0"> 
def VHLO_CustomCallOpV2 : VHLO_Op<"custom_call", [], "0.1.0"> 
```

The alternative is to make the trait list the last argument, in which case only ops that need traits have this ugliness, need to explicitly specify the default values to specify the traits:

```
def VHLO_ReturnOpV1 : VHLO_Op<"return", "0.0.0", "current", [Terminator]>
def VHLO_ReduceOpV1 : VHLO_Op<"reduce", "0.0.0", "current", [SameVariadicOperandSize]>
def VHLO_ScatterOpV1 : VHLO_Op<"scatter", "0.0.0", "current", [SameVariadicOperandSize]>
def VHLO_ReduceWindowOpV1 : VHLO_Op<"reduce_window", "0.0.0", "current", [SameVariadicOperandSize]> 
```

I'll go with the latter route for now since there are only those 4 ops that need traits currently. Removed `OpWithTraits` for now.

GleasonK commented on stablehlo/compatibility/transforms/MapStablehloToVersionedStablehlo.h:
```diff
@@ -0,0 +1,176 @@
+==============================================================================*/
+
+#ifndef MLIR_HLO_DIALECT_MHLO_TRANSFORMS_MAP_STABLEHLO_TO_HLO_OP_H
+#define MLIR_HLO_DIALECT_MHLO_TRANSFORMS_MAP_STABLEHLO_TO_HLO_OP_H
```
Fixed. Also found a few typos when trying to scan for other headers guards that may be misnamed (#668).

GleasonK replied:
Also had a thought regarding CI for this - I was having issues running clang-tidy on github CI because it tends to be flaky, however it may prove beneficial to run a more limited tidy CI. Maybe if we only checked header guards, and gradually built up from there depending on flakiness, we could get some useful CI in the near future.

burmako commented on stablehlo/transforms/LegalizeVhloToVersion.cpp:
```diff
@@ -0,0 +1,205 @@
+  VhloToVersionPass(VhloToVersionPassOptions const& opts)
+      : impl::VhloToVersionPassBase<VhloToVersionPass>(opts) {}
+
+  FailureOr<Version> parseTargetVersion(llvm::StringRef versionRef) {
```
Move helpers to free functions? Free functions are generally preferable to member functions because their lexical scope is smaller.

burmako commented on stablehlo/transforms/LegalizeVhloToVersion.cpp:
```diff
@@ -0,0 +1,205 @@
+    // An op is legal if the target version is in the ops `[min, max]`
+    // supported version range.
+    // Example:
+    //   CustomCallv1 0.0.0 -> 0.0.x
```
Capitalize "v" here and below? (Similarly to how op names are spelled right now).

burmako commented on stablehlo/dialect/VhloOps.td:
```diff
@@ -0,0 +1,937 @@
+  }];
+}
+
+def VHLO_ConstantOpV1 : VHLO_Op<"constant"> {
```
Let's also rename mnemonics to follow the op names? If this is a lot of work, let's defer the discussion and potential execution to a ticket.

GleasonK replied:
My initial design had `_v1` mnemonics and we decided to avoid that design because was more verbose without much benefit. The notion of un-suffixed operations being v1 exists in TF Lite, for example.

GleasonK commented on stablehlo/transforms/Passes.h:
```diff
@@ -0,0 +1,51 @@
+#include "mlir/Transforms/DialectConversion.h"
+
+namespace mlir {
+namespace vhlo {
```
Do we anticipate any other passes? The reason the file was originally named `CompatibilityPasses` was because all transformations of StableHLO were strictly for compatibility purposes, since we are a shallow opset with no optimizations.

If we think we'll eventually have other legalizations to other dialects (maybe to TCP if it gets moved to upstream eventually?) then I think it is right to call this `stablehlo`, otherwise would tend towards `vhlo` for the above reason.

GleasonK replied:
https://github.com/tensorflow/mlir-hlo/blob/0a819de96902c7a84a1fca6fb33abaded3569b6d/mhlo/transforms/rewriters.h#L182

Changed. Note this is a slight deviation from MHLO where the target dialect is the namespace used. We could split this so passes are in `stablehlo` and `populate` methods are in `vhlo`, but I'm not sure that subtlety is worth it.

GleasonK commented on stablehlo/dialect/VhloAttrs.td:
```diff
@@ -0,0 +1,172 @@
+limitations under the License.
+==============================================================================*/
+
+#ifndef STABLEHLO_COMPATIBILITY_VHLO_ATTRS
```
Ah shoot. I only scripted something that checked `.h` files. Will fix these too.

GleasonK commented on stablehlo/compatibility/dialect/VersionNumber.h:
```diff
@@ -0,0 +1,99 @@
+}  // namespace versionedhlo
+}  // namespace mlir
+
+#endif  // STABLEHLO_COMPATIBILITY_VERSION_NUMBER_H
```
Fixed. This would be a useful lint check as well. Made a note to add.

GleasonK commented on stablehlo/dialect/Version.h:
```diff
@@ -0,0 +1,92 @@
+      llvm_unreachable("failed to parse version number");
+    }
+    return num;
+  }
```
Done

GleasonK commented on stablehlo/dialect/CMakeLists.txt:
```diff
@@ -146,3 +146,29 @@ target_include_directories(StablehloOps INTERFACE
+  StablehloAssemblyFormat
+  MLIRIR
+  MLIRSupport
+  MLIRQuantDialect
```
I'm going to try to put some time aside to make some tooling for build in isolation. I have an experimental script in the works locally, but it doesn't seem to fail when I edit CMakeList files to have invalid dependencies, so need to explore a little further.

burmako commented on stablehlo/dialect/Version.h:
```diff
@@ -0,0 +1,66 @@
+
+  int64_t getMajorVersion() const { return majorMinorPatch[0]; }
+  int64_t getMinorVersion() const { return majorMinorPatch[1]; }
+  int64_t getPatchVersion() const { return majorMinorPatch[2]; }
```
Nitpick: getMajor/getMinor/getPatch given that we're inside `Version` already.

burmako commented on stablehlo/dialect/VhloOps.td:
```diff
@@ -0,0 +1,937 @@
+
+    Version log:
+      0.3.0: Bootstrap from MHLO: https://github.com/openxla/stablehlo/pull/1.
+      0.4.0: Add CustomCallOp input output aliasing: https://github.com/openxla/stablehlo/pull/403
```
Let's also batch other changes, as discussed in a nearby comment thread?

```
    Version log:
      0.3.0: Bootstrap from MHLO: https://github.com/openxla/stablehlo/pull/1.
      0.4.0: Add AllGatherOp::use_global_device_ids: https://github.com/openxla/stablehlo/pull/272.
             Add CollectivePermuteOp::channel_handle: https://github.com/openxla/stablehlo/pull/388.
             Add CustomCallOp::output_operand_aliases: https://github.com/openxla/stablehlo/pull/403.
```

GleasonK replied:
Split these VHLO ops into V1/V2 ops and added transformations / tests

burmako commented on stablehlo/transforms/LegalizeStablehloToVhlo.cpp:
```diff
@@ -0,0 +1,227 @@
+
+#define RETURN_CONVERTED_ENUM_ATTR(Name)                             \
+  auto stablehloValue = stablehlo::stringify##Name(attr.getValue()); \
+  auto hloValue = vhlo::symbolize##Name(stablehloValue);             \
```
`vhloValue`.

burmako commented on stablehlo/transforms/LegalizeStablehloToVhlo.cpp:
```diff
@@ -0,0 +1,227 @@
+  // with the exception of ArrayAttr which is converted recursively.
+  // This will change once we fork necessary upstream types to VHLO.
+  if (auto stablehloAttrs = stablehloAttr.dyn_cast<ArrayAttr>()) {
+    SmallVector<Attribute> hloAttrs;
```
`vhloAttrs`.

burmako commented on stablehlo/transforms/LegalizeStablehloToVhlo.cpp:
```diff
@@ -0,0 +1,227 @@
+  if (auto stablehloAttrs = stablehloAttr.dyn_cast<ArrayAttr>()) {
+    SmallVector<Attribute> hloAttrs;
+    for (auto stablehloAttr : stablehloAttrs) {
+      auto hloAttr = convertAttrToVhlo(stablehloAttr);
```
`vhloAttr`.

burmako commented on stablehlo/transforms/LegalizeStablehloToVhlo.cpp:
```diff
@@ -0,0 +1,227 @@
+          stablehloOp, vhloTypes, vhloOperands, vhloAttrs);
+    }
+
+    for (auto [hloRegion, stablehloRegion] :
```
`vhloRegion`.

burmako commented on stablehlo/transforms/LegalizeVhloToStablehlo.cpp:
```diff
@@ -0,0 +1,216 @@
+/// VHLO --> StableHLO ///
+//////////////////////////
+#define RETURN_CONVERTED_ENUM_ATTR(Name)                        \
+  auto stablehloValue = vhlo::stringify##Name(attr.getValue()); \
```
`vhloValue`.

burmako commented on stablehlo/transforms/LegalizeVhloToStablehlo.cpp:
```diff
@@ -0,0 +1,216 @@
+//////////////////////////
+#define RETURN_CONVERTED_ENUM_ATTR(Name)                        \
+  auto stablehloValue = vhlo::stringify##Name(attr.getValue()); \
+  auto hloValue = stablehlo::symbolize##Name(stablehloValue);   \
```
`stablehloValue`.

burmako commented on stablehlo/transforms/LegalizeVhloToStablehlo.cpp:
```diff
@@ -0,0 +1,216 @@
+  if (auto vhloAttrs = vhloAttr.dyn_cast<ArrayAttr>()) {
+    SmallVector<Attribute> hloAttrs;
+    for (auto vhloAttr : vhloAttrs) {
+      auto hloAttr = convertAttrToStablehlo(vhloAttr);
```
`stablehloAttr`.

burmako commented on stablehlo/transforms/LegalizeVhloToStablehlo.cpp:
```diff
@@ -0,0 +1,216 @@
+  // with the exception of ArrayAttr which is converted recursively.
+  // This will change once we fork necessary upstream types to VHLO.
+  if (auto vhloAttrs = vhloAttr.dyn_cast<ArrayAttr>()) {
+    SmallVector<Attribute> hloAttrs;
```
`stablehloAttrs`.

burmako commented on stablehlo/transforms/LegalizeVhloToStablehlo.cpp:
```diff
@@ -0,0 +1,216 @@
+          vhloOp, stablehloTypes, stablehloOperands, stablehloAttrs);
+    }
+
+    for (auto [hloRegion, stablehloRegion] :
```
`vhloRegion`.

burmako commented on stablehlo/transforms/VhloToVersion.cpp:
```diff
@@ -0,0 +1,211 @@
+///////////////////////
+namespace vhlo {
+namespace {
+using namespace vhlo;  // For VHLO Ops, Version, VhloDialect.
```
We are already in `namespace vhlo`, right?

GleasonK replied:
Whoops. I moved this all to `namespace vhlo` to explicitly get rid of the `using namespace` declaration. Had to shuffle things around to move passes to `namespace stablehlo`.

GleasonK commented:
Removed the `stablehlo_roundtrip.mlir` testing. This test was broken, and when I fixed it I realized that it wasn't the proper form of backward compatibility testing. We need a copy of this file per version of VHLO for testing backward compatibility. Otherwise it is only testing legalizations which is tested in the legalize testpoint.

burmako commented on stablehlo/dialect/VhloOps.td:
```diff
@@ -0,0 +1,957 @@
+        "::mlir::vhlo::CustomCallApiVersion::API_VERSION_ORIGINAL">:
+        $api_version,
+    DefaultValuedOptionalAttr<VHLO_FlatSymbolRefArrayAttr, "{}">:$called_computations,
+    OptionalAttr<VHLO_ArrayOfLayoutAttr>:$operand_layouts
```
This should also include result_layouts, and the patterns/tests will need to be updated.

GleasonK replied:
Fixed. OutputOperandAlias is a strange one because it is both Optional and DefaultValued.

I realized the policy for adding default valued attrs should be "if the value is the default value, we can downgrade," unless the default is a new behavior, which isn't the case here.

Meaning both of the following should be valid downgrades:

```
func.func @main(%arg0 : tensor<f32>) -> tensor<f32> {
  %0 = stablehlo.custom_call @foo(%arg0) {
    operand_layouts = [dense<> : tensor<0xindex>],
    result_layouts = [dense<> : tensor<0xindex>]
  } : (tensor<f32>) -> tensor<f32>
  %1 = stablehlo.custom_call @foo(%arg0) {
    operand_layouts = [dense<> : tensor<0xindex>],
    output_operand_aliases = [],
    result_layouts = [dense<> : tensor<0xindex>]
  } : (tensor<f32>) -> tensor<f32>
  func.return %0 : tensor<f32>
}
```

(one without output_operand_alias, one with the default value of output_operand_alias)

burmako reviewed (approved):
LGTM, except for one more comment.
